contracts/contracts/strategies/MorphoAaveStrategy.sol

```diff
@@ -0,0 +1,239 @@
+            address(_getPTokenFor(_asset)),
+            address(this), // the address of the user you want to supply on behalf of
+            _amount
+        );
+        emit Deposit(_asset, address(_getPTokenFor(_asset)), _amount);
+    }
+
+    /**
+     * @dev Deposit the entire balance of any supported asset into Morpho
+     */
+    function depositAll() external override onlyVault nonReentrant {
+        for (uint256 i = 0; i < assetsMapped.length; i++) {
+            uint256 balance = IERC20(assetsMapped[i]).balanceOf(address(this));
+            if (balance > 0) {
+                _deposit(assetsMapped[i], balance);
+            }
+        }
+    }
+
+    /**
+     * @dev Withdraw asset from Morpho
+     * @param _recipient Address to receive withdrawn asset
+     * @param _asset Address of asset to withdraw
+     * @param _amount Amount of asset to withdraw
+     */
+    function withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) external override onlyVault nonReentrant {
+        _withdraw(_recipient, _asset, _amount);
+    }
+
+    function _withdraw(
+        address _recipient,
+        address _asset,
+        uint256 _amount
+    ) internal {
+        require(_amount > 0, "Must withdraw something");
+        require(_recipient != address(0), "Must specify recipient");
+
+        address pToken = assetToPToken[_asset];
+
+        IMorpho(MORPHO).withdraw(pToken, _amount);
+        emit Withdrawal(_asset, address(_getPTokenFor(_asset)), _amount);
+        IERC20(_asset).safeTransfer(_recipient, _amount);
+    }
+
+    /**
+     * @dev Remove all assets from platform and send them to Vault contract.
+     */
+    function withdrawAll() external override onlyVaultOrGovernor nonReentrant {
+        for (uint256 i = 0; i < assetsMapped.length; i++) {
+            uint256 balance = _checkBalance(assetsMapped[i]);
+            if (balance > 0) {
+                _withdraw(vaultAddress, assetsMapped[i], balance);
+            }
+        }
+    }
+
+    /**
+     * @dev Return total value of an asset held in the platform
+     * @param _asset      Address of the asset
+     * @return balance    Total value of the asset in the platform
+     */
+    function checkBalance(address _asset)
+        external
+        view
+        override
+        returns (uint256 balance)
+    {
+        return _checkBalance(_asset);
+    }
+
+    function _checkBalance(address _asset)
+        internal
+        view
+        returns (uint256 balance)
+    {
+        address pToken = assetToPToken[_asset];
+
+        // Total value represented by decimal position of underlying token
+        (, , balance) = ILens(LENS).getCurrentSupplyBalanceInOf(
+            pToken,
+            address(this)
+        );
+    }
+
+    /**
+     * @dev Retuns bool indicating whether asset is supported by strategy
+     * @param _asset Address of the asset
+     */
+    function supportsAsset(address _asset)
```

Comment:
[MorphoCompound strategy](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/strategies/MorphoCompoundStrategy.sol#L17) extends [BaseCompoundStrategy](https://github.com/OriginProtocol/origin-dollar/blob/master/contracts/contracts/strategies/BaseCompoundStrategy.sol#L23-L42) that basically has  `supportsAsset` and `_getP(C)TokenFor ` already implemented. 

You could do the same thing here, and not needing to implement 2 additional functions.

But I get it that base contract is Compound and not Aave...

I wonder if we should rename `_getCTokenFor` to `_getPTokenFor` in `BaseCompoundStrategy` and move it to `InitializableAbstractStrategy`. There is some code duplication that might be cool to clean up... though not having a strong opinion on this one

Reply:
I actually started off by inheriting the `BaseCompoundStrategy` but the `cToken` nomenclature and some compound-specific functions (like `_convertUnderlyingToCToken`) were what caused me to remove that inheritance.

But I agree that we should avoid code redundancy of some methods like `supportsAsset` and `_getPTokenFor`. I guess we could have a new abstract contract `BaseGenericStrategy` (ik, I'm bad at naming :P) or something, which `BaseCompoundStrategy` and these Morpho strategies can inherit from. Let me explore what options we have

Reply:
I think we would be getting into the realm of too much inheritance. The methods duplicated in this contract are small and focused. I don't think there's harm in it. Security, and via extension, understandability are are the biggest things we need to prioritize. Flatter contracts trees are much easier to review and understand, and sometimes some duplication is a worthwhile tradeoff for what's really important.

Reply:
Yep, totally agree with that

Reply:
agree, sounds good to me